packages/client/composables/useDragElements.ts

```diff
@@ -0,0 +1,71 @@
+import { debounce } from '@antfu/utils'
+import { useDynamicSlideInfo } from './useSlideInfo'
+
+export interface DragElementsContext {
+  register: (id: string) => void
+  unregister: (id: string) => void
+  update: (id: string, posStr: string) => void
+  save: () => Promise<void>
+}
+
+const map: Record<number, DragElementsContext> = {}
+
+export function useDragElementsContext(no: number): DragElementsContext {
+  if (!(__DEV__ && __SLIDEV_FEATURE_EDITOR__)) {
+    return {
+      register() { },
+      unregister() { },
+      update() { },
+      save: async () => { },
+    }
+  }
+
+  if (map[no])
+    return map[no]
+  const { info, update } = useDynamicSlideInfo(no)
+
+  const elements: string[] = []
+
+  let newContent: string | null = null
+  async function save() {
+    if (newContent) {
+      await update({
+        content: newContent,
+        skipHmr: true,
+      })
+      newContent = null
+    }
+  }
+  const debouncedSave = debounce(500, save)
+
+  return map[no] = {
+    register(id) {
+      elements.push(id)
+    },
+    unregister(id) {
+      elements.splice(elements.indexOf(id), 1)
+    },
+    update(id, posStr) {
+      const idx = elements.indexOf(id)
+      if (idx < 0 || !info.value)
+        return
+      const oldContent = info.value.content
+      const match = [...oldContent.matchAll(/<(v-?drag)(.*?)(?:pos=".*?")(.*?)>/ig)][idx]
```

Comment:
I guess it would be better to move these codemod logic to the server side.

Reply:
Also, I think it might be better that we store the position in the frontmatter, that we could also assign IDs to v-drag and resolve the pos at runtime. For example:

```md
---
dragPos:
  foo: -8,134,153,48
---

<v-drag drag-id="foo">
foo
</v-drag>
```

This way we would be able to do positions even for `v-drag` inside components or some tricky places that regex couldn't cover.

Reply:
> Also, I think it might be better that we store the position in the frontmatter, that we could also assign IDs to v-drag and resolve the pos at runtime. For example:

I am wondering if this is the most convenient way. Users have to specify the ID by themselves, which can be quite annoying in my opinion. If we auto-generate IDs if no ID is provided, we have to use Regex again and the auto-generated IDs are not pretty. Since it is quite rare to use `v-drag` inside a component or in the code block, maybe we could give up this support as a trade-off.

Reply:
I am thinking the most robust and generalized way, while you are considering the most user-friendly way - I think both are indeed valid perspectives.

The reason I lean more to robust is for long-term maintenance, the approach we are taken right now would lead to many false-positives we need to handle in the future. Also accidentally write positions to the wrong element might be annoying an confusing to users.

I guess we could support both - `pos` for in markdown notations, and `id` for v-drag in components, that we could throw errors when we failed to find the notation and ask users to explicit provide id when we failed to resolve the data, etc.

Also think we might be able to use a markdown-it plugin to track the source position more accurately.

Reply:
I've just renamed the types `DragElements***` to `DragElement***`, and allow users to neither specify `id` nor `pos` at the initial.

Reply:
btw, I am going to implement `height: unset` tomorrow, which is quite useful I think.